distributed/scheduler.py

```diff
@@ -1823,17 +1823,21 @@ def update_graph(
                     self.report({"op": "cancelled-key", "key": k}, client=client)
                     self.client_releases_keys(keys=[k], client=client)
 
+        # Avoid computation that is already finished
+        already_in_memory = set()  # tasks that are already done
+
         # Remove any self-dependencies (happens on test_publish_bag() and others)
         for k, v in dependencies.items():
             deps = set(v)
             if k in deps:
                 deps.remove(k)
             dependencies[k] = deps
 
-        # Avoid computation that is already finished
-        already_in_memory = set()  # tasks that are already done
-        for k, v in dependencies.items():
-            if v and k in self.tasks and self.tasks[k].state in ("memory", "erred"):
+            if (
+                dependencies[k]
```

Comment:
You already have this as `v` from the loop I believe.
```suggestion
                v
```